File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/armor/GenerationsArmorItem.java
package generations.gg.generations.core.generationscore.common.world.item.armor;

import com.google.common.collect.ImmutableMultimap;
import com.google.common.collect.Multimap;
import dev.architectury.extensions.ItemExtension;
import net.minecraft.Util;
import net.minecraft.core.Holder;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.entity.ai.attributes.Attribute;
import net.minecraft.world.entity.ai.attributes.AttributeModifier;
import net.minecraft.world.entity.ai.attributes.Attributes;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.*;
import net.minecraft.world.item.component.ItemAttributeModifiers;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class GenerationsArmorItem extends ArmorItem implements ItemExtension {
    public final Set<ArmorTickEffect> armorEffects = new HashSet<>();
    public final Set<CustomAttributeModifier> customAttributeModifiers = new HashSet<>();

    public GenerationsArmorItem(Holder<ArmorMaterial> armorMaterial, Type equipmentSlot, Properties properties, ArmorEffect... armorEffects) {
        super(armorMaterial, equipmentSlot, properties);

        for (ArmorEffect armorEffect : armorEffects) {
            if (armorEffect instanceof ArmorTickEffect effect) {
                this.armorEffects.add(effect);
            } else if (armorEffect instanceof CustomAttributeModifier effect) {
                this.customAttributeModifiers.add(effect);
            }
        }
    }


//    TODO: Figure out wtf I'm gonna do with mess being upended.
//    @Override
//    public ItemAttributeModifiers getDefaultAttributeModifiers() {
//        var entries = customAttributeModifiers.stream().map()
//
//        return super.getDefaultAttributeModifiers();
//    }
//
//    //Method for Forge
//    public Multimap<Attribute, AttributeModifier> getAttributeModifiers(EquipmentSlot equipmentSlot, ItemStack itemStack) {
//        var builder = ImmutableMultimap.<Attribute, AttributeModifier>builder();
//        for (CustomAttributeModifier customAttributeModifier : customAttributeModifiers) {
//            customAttributeModifier.getAttributeModifiers(builder, equipmentSlot, itemStack, this);
//        }
//
//        builder.putAll(this.getDefaultAttributeModifiers(equipmentSlot));
//
//        return builder.build();
//    }
//
//    //Method for Fabric
//    public Multimap<Attribute, AttributeModifier> getAttributeModifiers(ItemStack stack, EquipmentSlot slot) {
//        return getAttributeModifiers(slot, stack);
//    }
//
//    @Override
//    public @NotNull Multimap<Attribute, AttributeModifier> getDefaultAttributeModifiers(@NotNull EquipmentSlot slot) {
//        return super.getDefaultAttributeModifiers(slot);
//    }

    @Override
    public void inventoryTick(@NotNull ItemStack itemStack, @NotNull Level world, @NotNull Entity entity, int slotId, boolean isSelected) {
        this.armorEffects.forEach(armorEffect -> armorEffect.inventoryTick(itemStack, world, entity, slotId, isSelected, this));
    }

    @Override
    public void tickArmor(ItemStack stack, Player player) {
        this.armorEffects.forEach(armorEffect -> armorEffect.onArmorTick(stack, player.level(), player, this));
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/item/armor/GenerationsArmorItem.kt
package generations.gg.generations.core.generationscore.common.world.item.armor

import com.google.common.base.Supplier
import com.google.common.base.Suppliers
import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
import net.minecraft.core.Holder
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EquipmentSlotGroup
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.ai.attributes.Attribute
import net.minecraft.world.entity.ai.attributes.AttributeModifier
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.item.ArmorItem
import net.minecraft.world.item.ArmorMaterial
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.ItemAttributeModifiers
import net.minecraft.world.level.Level
import java.util.function.Consumer

class GenerationsArmorItem(
    armorMaterial: Holder<ArmorMaterial>,
    equipmentSlot: Type,
    properties: Properties,
    additionalModfier: List<Pair<Holder<Attribute>, AttributeModifier>>,
    vararg armorEffects: ArmorEffect?,
) :
    ArmorItem(armorMaterial, equipmentSlot, properties) {
    private var attributeModifiers: Supplier<ItemAttributeModifiers> = Suppliers.memoize<ItemAttributeModifiers>(Supplier {
        val i = (material.value() as ArmorMaterial).getDefense(type)
        val f = (material.value() as ArmorMaterial).toughness()
        val builder = ItemAttributeModifiers.builder()
        val equipmentSlotGroup = EquipmentSlotGroup.bySlot(type.getSlot())
        val resourceLocation = ResourceLocation.withDefaultNamespace("armor." + type.getName())
        builder.add(
            Attributes.ARMOR,
            AttributeModifier(resourceLocation, i.toDouble(), AttributeModifier.Operation.ADD_VALUE),
            equipmentSlotGroup
        )
        builder.add(
            Attributes.ARMOR_TOUGHNESS,
            AttributeModifier(resourceLocation, f.toDouble(), AttributeModifier.Operation.ADD_VALUE),
            equipmentSlotGroup
        )
        val g = (material.value() as ArmorMaterial).knockbackResistance()
        if (g > 0.0f) {
            builder.add(
                Attributes.KNOCKBACK_RESISTANCE,
                AttributeModifier(resourceLocation, g.toDouble(), AttributeModifier.Operation.ADD_VALUE),
                equipmentSlotGroup
            )
        }

        additionalModfier.forEach {

            builder.add(it.first, it.second, EquipmentSlotGroup.bySlot(type.slot))}

        builder.build()
    })

    val armorEffects: MutableSet<ArmorTickEffect> = HashSet()
    val customAttributeModifiers: MutableSet<CustomAttributeModifier> = HashSet()

    init {

        for (armorEffect in armorEffects) {
            if (armorEffect is ArmorTickEffect) {
                this.armorEffects.add(armorEffect)
            } else if (armorEffect is CustomAttributeModifier) {
                customAttributeModifiers.add(armorEffect)
            }
        }
    }


    //    TODO: Figure out wtf I'm gonna do with mess being upended.
    //    @Override
    //    public ItemAttributeModifiers getDefaultAttributeModifiers() {
    //        var entries = customAttributeModifiers.stream().map()
    //
    //        return super.getDefaultAttributeModifiers();
    //    }
    //
    //    //Method for Forge
    //    public Multimap<Attribute, AttributeModifier> getAttributeModifiers(EquipmentSlot equipmentSlot, ItemStack itemStack) {
    //        var builder = ImmutableMultimap.<Attribute, AttributeModifier>builder();
    //        for (CustomAttributeModifier customAttributeModifier : customAttributeModifiers) {
    //            customAttributeModifier.getAttributeModifiers(builder, equipmentSlot, itemStack, this);
    //        }
    //
    //        builder.putAll(this.getDefaultAttributeModifiers(equipmentSlot));
    //
    //        return builder.build();
    //    }
    //
    //    //Method for Fabric
    //    public Multimap<Attribute, AttributeModifier> getAttributeModifiers(ItemStack stack, EquipmentSlot slot) {
    //        return getAttributeModifiers(slot, stack);
    //    }
    //
    //    @Override
    //    public @NotNull Multimap<Attribute, AttributeModifier> getDefaultAttributeModifiers(@NotNull EquipmentSlot slot) {
    //        return super.getDefaultAttributeModifiers(slot);
    //    }
    override fun inventoryTick(itemStack: ItemStack, world: Level, entity: Entity, slotId: Int, isSelected: Boolean) {
        val livingEntity = entity.instanceOrNull<LivingEntity>() ?: return

        armorEffects.forEach(Consumer { armorEffect: ArmorTickEffect ->
            armorEffect.inventoryTick(
                itemStack, world, livingEntity, slotId, isSelected,
                this
            )
        })
    }

    override fun getDefaultAttributeModifiers(): ItemAttributeModifiers? {
        return attributeModifiers.get();
    }
}
